PlonkUp

davidnevadoc opened the pull request:
The purpose of this PR is to include lookup gates to achieve a combination of the original Plonk + Plookup (known as PlonkUp).
Closes #41

LukePearson1 committed:
Port lookup tables

LukePearson1 committed:
Port multisets

LukePearson1 committed:
Port multisets

LukePearson1 committed:
Fix compilation

LukePearson1 committed:
Implement lookups

LukePearson1 committed:
Restore files

LukePearson1 committed:
Apply cargo fmt

LukePearson1 committed:
Import polynomial

LukePearson1 committed:
Refactor multiset

LukePearson1 committed:
Add lookups

lopeetall committed:
Added lookup table tests for both fields and fmt

lopeetall committed:
Added multiset tests for both fields and fmt

lopeetall committed:
Added witness table tests for both fields

lopeetall committed:
Added table preprocessing test for both fields

LukePearson1 committed:
Include Rand crate

LukePearson1 committed:
Refactor prover to include lookups (#84)

* Refactor prover to include lookups

This is the first commit to implement the
plonkup proving system.

* Reference Scalar

* Lookup error fixing (#87)

* Include PairingEngine in impls calling ProverKey

* Added lookup fields and derefed where needed

* Fixed some poly math

Co-authored-by: Joshua Fitzgerald <joshuabfitzgerald@gmail.com>

* append challenges to transcript

* assert challenges  different

* compute permutation challenges

* add missing challenges commitments

* Fix referenced variables

* Partial refactor

This commit contains various fixes
to the proving algorithm to implement
trait based features

* Remove some errors

* Refactor proving algorithm

* Finalise prover algorithm

* removed unneeded q_lookup padding; fixed some comments

Co-authored-by: Joshua Fitzgerald <joshuabfitzgerald@gmail.com>
Co-authored-by: mathcrypto <bouguera_amira@yahoo.fr>

lopeetall committed:
Remove more dup imports

lopeetall committed:
comment out lookup contributions

lopeetall committed:
ProverKey<F> -> ProverKey<F, PC> for table commits in lookup proverkey

add trait bounds for canonical ser/de

lopeetall committed:
append_scalar -> append

lopeetall committed:
fix multiset sorted halve

davidnevadoc committed:
Cargo fmt

davidnevadoc committed:
Add linear combination utility function

davidnevadoc committed:
Clean up linear combinations in `LookUp` widget

davidnevadoc committed:
Add missing lookup selector to `ProverKey`

davidnevadoc committed:
Remove `PC` from `ProverKey`

Some extra lines in diff due to formatting issues

davidnevadoc committed:
Remove outdated blinding strategy for lookups

davidnevadoc committed:
Remove `PC` from quot and lin polys computation

davidnevadoc committed:
Fix `lc` function

davidnevadoc committed:
Remove duplicated files

davidnevadoc committed:
Refactor `MultiSet`

Improve `compress` function + minor fixes

lopeetall committed:
Change to new compression function

lopeetall committed:
add table columns to prover preprocessing and serilization

lopeetall committed:
Adjust typing in table preprocessing

lopeetall committed:
Fix sorted halve

lopeetall committed:
Fix test_table_preprocessing

lopeetall committed:
Fix compress function

lopeetall committed:
Add missing docs for lookups

lopeetall committed:
Fmt fixup lookup/preprocessing

lopeetall committed:
Fixup fmt add table columns

lopeetall committed:
Small error fixes

davidnevadoc committed:
Add generic `PC` in table preprocessing

Remove KZG commitments + fix tests

davidnevadoc committed:
Merge branch 'PlookUp' of github.com:rust-zkp/ark-plonk into PlookUp

davidnevadoc committed:
Replace repeated code with loops

davidnevadoc committed:
Fix `src/lib.rs`

Old module organization removed

davidnevadoc committed:
Cargo fix

 - Remove unused imports
 - Mark unused variables

lopeetall committed:
Include q_lookup in add_blinding_factors

davidnevadoc committed:
Add `push` and `len` to `LookupTable`

davidnevadoc committed:
Remove repeated code in `WitnessTable`

davidnevadoc committed:
Small fixes

davidnevadoc committed:
Remove std feature

davidnevadoc committed:
Fix `sorted_halve` function + empty table bug fix

davidnevadoc committed:
Fix `WitnessTable` bug

davidnevadoc committed:
Implement `sorted_concat`

Unoptimized first version

davidnevadoc committed:
Rewrite `f_multiset` computation

lopeetall committed:
Add simple plookup test

lopeetall committed:
Plook up separate lookups (#124)

* restore original permutation.rs

* update linearisation with lookups

* make first lagrange eval fn public

* restore original quotient_poly.rs

* add new lookup.rs

* add lookup to quotient poly

* separate lookup quotient evaluation

* move, tidy, and fmt

* Add lookup contribution to `VerifierKey`

* `theta` is not `zeta`

* revert some documentation to original

* add table column commitments to lookup verifier key

* move table preprocessing to preprocess_shared

* simpler cast domain.size() to u32

Co-authored-by: David Nevado <david.nevado@iohk.io>

lopeetall committed:
add a second gate to plookup xor test

lopeetall committed:
fix sorted_halve to avoid absolute sorting

lopeetall committed:
fix sorted_halve to remove dupes

lopeetall committed:
Plookup passing (#125)

* fix first lagrange eval computation

* copy lookup debug changes: h1 -> h1_next, clean up lookup computations

* sync transcripts

* add pretty printers to util

* debug: remove blinding factors

* debug: fix challenges to constants, print multisets

* debug: quotient poly debug mode

* debug: match quotient poly debug mode in lookup widget

* debug: rename composer debug printer

* btreemap -> hashmap

* fix sorted_halve

* debug: allow challenges to vary

* debug: enable more aw and saw checks

* debug: turn off quotient debug mode

* remove debuggers

* add new sorted_halve

* add table commitments

* turn blinding back on

* remove leftover debug functions

* restore reference to quotient poly

* implement clippy suggestions

* tag lookup len() function as unused to please CI

* fix missing q_lookup and duped q_fixed_group_add pushes

* make sure q_lookup is padded before constucting f_scalars

davidnevadoc committed:
Add error to `sorted_halve` function

lopeetall committed:
make plookup xor test more robust

lopeetall committed:
enable padding when lookup table is larger than circuit

davidnevadoc committed:
Minor fixes

weikengchen commented on plonk-core/Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "plonk-core"
-version = "0.8.2"
+version = "0.0.0"
```
Why?

CPerezz replied:
Because we need to re-ship the code. Which was originally a fork of `dusk-plonk`. But  we are that far away from them and have much more improvements that we want to start publishing to crates with a clean version.

weikengchen commented on plonk-core/src/constraint_system/composer.rs:
```diff
@@ -17,11 +17,13 @@
 use core::marker::PhantomData;
 use hashbrown::HashMap;
 use rand::{CryptoRng, RngCore};
+use std::cmp::max;
```
```suggestion
use ark_std::cmp::max;
```

A `cargo fmt` will be needed after this change.

bhgomes replied:
Please use `core::cmp::max` for this.

weikengchen replied:
So in ZK-Garage we prefer core::/alloc:: instead of ark_std::? Just to follow the habits here.

Also in arkworks, we generally get rand:: from ark_std::rand:: to avoid version conflicts.

bhgomes replied:
We should just be using `rand_core` anyway, which has a very stable interface and is actually `no_std`.

davidnevadoc replied:
Switched to `core::cmp::max` and `rand` has been substituted for `rand_core`.

bhgomes commented on plonk-core/Cargo.toml:
```diff
@@ -64,6 +64,7 @@ merlin = { version = "3.0", default-features = false }
 num-traits = { version = "0.2.14" }
 rand = { version = "0.8", default-features = false, features = ["getrandom"] }
 
+
```
nit: remove whitespace

bhgomes commented on plonk-core/src/constraint_system/composer.rs:
```diff
@@ -122,6 +129,11 @@ where
+    /// Returns the length of the circuit that can accomodate the lookup table
+    pub fn total_size(&self) -> usize {
+        max(self.n, self.lookup_table.size())
+    }
```
Because we now have two size measuring functions, it could be potentially confusing when getting the correct circuit bound. We should just add the upper bound method to the `StandardComposer`:

```rust
pub fn circuit_bound(&self) -> usize {
    self.total_size().next_power_of_two()
}
```

feel free to bikeshed the name.

davidnevadoc replied:
Good point! I have added `circuit_bound` as the only method available to avoid confusion.

bhgomes commented on plonk-core/src/constraint_system/helper.rs:
```diff
@@ -63,7 +63,7 @@ where
         let (ck, _) = PC::trim(
             &universal_params,
-            prover.circuit_size().next_power_of_two(),
+            prover.total_size().next_power_of_two(),
```
Use `StandardComposer::circuit_bound`.

bhgomes commented on plonk-core/src/constraint_system/helper.rs:
```diff
@@ -93,7 +93,7 @@ where
     let (ck, vk) = PC::trim(
         &universal_params,
-        verifier.circuit_size().next_power_of_two(),
+        verifier.total_size().next_power_of_two(),
```
Use `StandardComposer::circuit_bound`.

bhgomes commented on plonk-core/src/constraint_system/lookup.rs:
```diff
@@ -0,0 +1,191 @@
+
+                let rand1 = rng.next_u64() % 16;
+                let rand2 = rng.next_u64() % 16;
+                let rand3 = rng.next_u64() % 16;
```
You can just use `RngCore::next_u32` since you are sampling `[0,16)`.

bhgomes commented on plonk-core/src/proof_system/preprocess.rs:
```diff
@@ -359,12 +384,17 @@ where
+            commitments[12].commitment().clone(), // left_sigma_poly_commit.0,
+            commitments[13].commitment().clone(), // right_sigma_poly_commit.0,
+            commitments[14].commitment().clone(), // out_sigma_poly_commit.0,
+            commitments[15].commitment().clone(), /* fourth_sigma_poly_commit.0, */
```
nit: use consistent comments

bhgomes commented on src/lib.rs:
```diff
@@ -10,7 +10,6 @@
 #![cfg_attr(doc_cfg, feature(doc_cfg))]
 #![forbid(rustdoc::broken_intra_doc_links)]
 #![forbid(missing_docs)]
-
```
nit: keep whitespace

bhgomes commented on plonk-core/src/util.rs:
```diff
@@ -143,6 +144,24 @@ where
+
+    values
+        .iter()
+        .rev()
```
Why is this reversed?

davidnevadoc replied:
In order to compute `v0 + challenge * v1 + challenge^2 * v2 + ... + challenge^k * vk` using k-1 mutliplications we need to iterate through the values from `vk` to `v0` so the order has to be reversed.

bhgomes commented on plonk-core/src/util.rs:
```diff
@@ -143,6 +144,24 @@ where
+///
+/// For values [v_0, v_1,... v_k] returns:
+/// v_0 + challenge * v_1 + ... + challenge^k  * v_k
+pub fn lc<F>(values: Vec<F>, challenge: F) -> F
```
`values` should just be passed in as `&[F]`, and `challenge` as `&F`

bhgomes commented on plonk-core/src/lookup/multiset.rs:
```diff
@@ -0,0 +1,525 @@
+        if n > self.len() as u32 {
+            let diff = n - self.len() as u32;
+            self.0.extend(vec![self.0[0]; diff as usize]);
+        }
```
I suggest the following:

```rust
let n = n as usize;
if n > self.len() {
    self.0.resize(n, self.0[0])
}
```

to replace those four lines.

bhgomes commented on plonk-core/src/lookup/multiset.rs:
```diff
@@ -0,0 +1,525 @@
+        self.0.into_iter().map(|x| x * elem).collect()
+    }
+}
+#[cfg(test)]
```
nit: add whitespace

bhgomes commented on plonk-core/src/lookup/multiset.rs:
```diff
@@ -0,0 +1,525 @@
+
+    /// Concatenates with a new `MultiSet` and sort
+    pub fn sorted_concat(&mut self, other: Self) {
+        // TODO This could be a performance bottleneck
```
Is this method ever even used?

lopeetall replied:
removed `sorted_concat`

bhgomes commented on plonk-core/src/lookup/multiset.rs:
```diff
@@ -0,0 +1,525 @@
+        multisets
+            .iter()
+            .rev()
+            .fold(MultiSet::with_len(len), |acc, m| acc * alpha + m.clone())
```
Should extend the `util::lc` to support this. Just add more generics and you can get this impl for free.

bhgomes commented on plonk-core/src/lookup/multiset.rs:
```diff
@@ -0,0 +1,525 @@
+    /// We combine the multisets together and sort their elements together.
+    /// The final MultiSet will look as follows, s: {1,1,2,2,3,3,4,4}.
+    /// Splits a multiset into alternating halves of the same length (if even).
+    pub fn sorted_halve(&self, f: &Self) -> Result<(Self, Self), Error> {
```
This method does not actually sort anything, the name and documentation are misleading. Rather, it is grouping the field points into buckets of the same value and concatenating the buckets, then splitting them into evens and odds.

bhgomes commented on plonk-core/src/lookup/multiset.rs:
```diff
@@ -0,0 +1,525 @@
+                _ => {
+                    counters.insert(*element, 1);
+                }
+            };
```
nit: remove `;`

lopeetall replied:
`insert` returns an Option while the other case returns `()`, so the compiler complains about incompatible types in the match without the semi-colon. Is there a better way to resolve this?

bhgomes replied:
I'm saying remove the semicolon at the bottom of the match, not the one in the `_` case, that one is required.

bhgomes commented on plonk-core/src/lookup/multiset.rs:
```diff
@@ -0,0 +1,525 @@
+    pub fn sorted_halve(&self, f: &Self) -> Result<(Self, Self), Error> {
+        let mut counters: HashMap<F, usize> = HashMap::new();
+
+        // Insert elemnts on of t in sorted struct
```
Comment misleading, remove

lopeetall replied:
rewritten

bhgomes commented on plonk-core/src/lookup/multiset.rs:
```diff
@@ -0,0 +1,525 @@
+            };
+        }
+
+        // Insert elements on of f in sorted struct + check they are in t
```
Comment misleading, remove

lopeetall replied:
rewritten

bhgomes commented on plonk-core/src/lookup/multiset.rs:
```diff
@@ -0,0 +1,525 @@
+                }
+            }
+            *count = 0usize;
+        });
```
I suggest the following loop instead:

```rust
for (elem, count) in counters {
    let half_count = count / 2;
    evens.extend(vec![elem; half_count]);
    odds.extend(vec![elem; half_count]);
    if count % 2 == 1 {
        if parity == 1 {
            odds.push(elem);
            parity = 0;
        } else {
            evens.push(elem);
            parity = 1;
        }
    }
}
```

lopeetall replied:
This loop creates an iterator out of the HashMap `counters` which automatically performs a sort. To avoid the sort and preserve the ordering in `t` we iterate over `t` instead in the original loop.

bhgomes replied:
I agree that iterating over the map doesn't preserve the ordering of `t`, (and so we should iterate over `self.0` instead), but I don't agree that it sorts anything. The `HashMap` iteration order is arbitrary (determined by the hash values), not sorted. See: https://docs.rs/hashbrown/latest/hashbrown/hash_map/struct.HashMap.html#method.into_iter-2

bhgomes replied:
In this case though, I still think `for` loop is preferable:

```rust
for elem in &self.0 {
    let count = counters.get_mut(elem).unwrap();
    let half_count = count / 2;
    evens.extend(vec![elem; half_count]);
    odds.extend(vec![elem; half_count]);
    if count % 2 == 1 {
        if parity == 1 {
            odds.push(elem);
            parity = 0;
        } else {
            evens.push(elem);
            parity = 1;
        }
    }
    *count = 0;
}
```

but if we use an `indexmap::IndexMap` instead of a `hashbrown::HashMap`, then we can use the first snippet I proposed because it preserves insert order (and its `no_std`) and it would reduce the number of loops we perform (as well as clearing the map as we go so we reduce the amount of space we use).

lopeetall replied:
Ah, you're right. It doesn't sort, it just changes the order.

Nice find with `IndexMap`! I'll change over to `IndexMap` and test with your original `for` loop.

lopeetall replied:
Works perfectly.

bhgomes commented on plonk-core/src/lookup/multiset.rs:
```diff
@@ -0,0 +1,525 @@
+    /// This function will be used to check if the all elements
+    /// in set f, from the paper, are contained inside t.
+    pub fn contains_all(&self, other: &Self) -> bool {
+        // Unoptimized function only used for testing
```
Move this comment to the actual doc string or mark this function as `pub(crate)`.

lopeetall replied:
Did both and included `#[allow(dead_code)]`

bhgomes commented on plonk-core/src/lookup/multiset.rs:
```diff
@@ -0,0 +1,525 @@
+            }
+        }
+        (Self(evens), Self(odds))
+    }
```
You should return `Option<(Self, Self)>` so that we capture the case that it is not even cardinality. If "evenness" is known at compile-time, then use a different type called `EvenCardinalityMultiSet` or something, and put the halving function on there.

Also, this function is not even used anywhere.

lopeetall replied:
removed `halve_alternating`

bhgomes commented on plonk-core/src/lookup/multiset.rs:
```diff
@@ -0,0 +1,525 @@
+    /// The last element of the first half should be the same
+    /// as the first element of the second half.
+    /// Since a multiset can never have an even cardinality, we
+    /// always split it in the way described above.
```
This docstring is in direct conflict with the one for `MultiSet::halve_alternating`. And this method is never used anyway.

lopeetall replied:
removed `halve_alternating`

bhgomes commented on plonk-core/src/lookup/multiset.rs:
```diff
@@ -0,0 +1,525 @@
+    use ark_poly::EvaluationDomain;
+    use ark_poly::Polynomial;
+
+    // FIXME: Run tests on both BLS fields.
```
resolved

lopeetall replied:
removed

bhgomes commented on plonk-core/src/lookup/preprocess.rs:
```diff
@@ -0,0 +1,147 @@
+                })
+                .collect::<Result<_, Error>>()?;
+
+        Ok(Self { n, t: result })
```
This can be simplified to this:

```rust
        assert!(n.is_power_of_two());
        let domain = EvaluationDomain::new(n as usize).unwrap();
        let result = table
            .vec_to_multiset()
            .into_iter()
            .enumerate()
            .map(|(index, mut column)| {
                column.pad(n);
                let poly = column.to_polynomial(&domain);
                let poly_name = format!("t_{}_poly", index + 1);
                let labeled_poly = ark_poly_commit::LabeledPolynomial::new(
                    poly_name,
                    poly.to_owned(),
                    None,
                    None,
                );
                let commitment = PC::commit(commit_key, &[labeled_poly], None)
                    .map_err(to_pc_error::<F, PC>)?;
                Ok((column, commitment.0[0].commitment().clone(), poly))
            })
            .collect::<Result<_, Error>>()?;
        Ok(Self { n, t: result })
```

which removes automatically inferred types, and uses a more canonical power of two check

bhgomes commented on plonk-core/src/permutation/mod.rs:
```diff
@@ -751,10 +751,97 @@ impl Permutation {
-/// We wil use this helper function to compensate.
+/// We will use this helper function to compensate.
 #[allow(dead_code)]
 pub(crate) fn random_scalar<F: FftField, R: RngCore>(rng: &mut R) -> F {
```
This function is never used, remove.

bhgomes commented on plonk-core/src/permutation/mod.rs:
```diff
@@ -751,10 +751,97 @@ impl Permutation {
+        assert_eq!(f.len(), domain.size());
+        assert_eq!(t.len(), domain.size());
+        assert_eq!(h_1.len(), domain.size());
+        assert_eq!(h_2.len(), domain.size());
```
Use `n` in these assertions.

bhgomes commented on plonk-core/src/permutation/mod.rs:
```diff
@@ -751,10 +751,97 @@ impl Permutation {
+                    ),
+                )
+            })
+            .map(|(num, den)| num * den.inverse().unwrap())
```
Combine the two maps into one.

bhgomes commented on plonk-core/src/permutation/mod.rs:
```diff
@@ -751,10 +751,97 @@ impl Permutation {
+        // remove the last element
+        p.remove(n);
+
+        assert_eq!(n, p.len());
```
I would recommend the following implementation:

```rust
let mut state = F::one();
let mut p = Vec::with_capacity(n + 1);
p.push(state);
for s in product_arguments {
    state *= s;
    p.push(state);
}
p.pop();
assert_eq!(n, p.len());
```

bhgomes commented on plonk-core/src/permutation/mod.rs:
```diff
@@ -751,10 +751,97 @@ impl Permutation {
+        let prod_2 = epsilon_plus_one_delta + h_2 + (h_1_next * delta);
+
+        prod_1 * prod_2
+    }
```
You can reduce the number of operations here:

```rust
fn lookup_ratio(delta: F, epsilon: F, f: F, t: F, t_next: F, h_1: F, h_1_next: F, h_2: F) -> (F, F) {
    let one_plus_delta = F::one() + delta;
    let epsilon_one_plus_delta = epsilon * one_plus_delta;
    (
        one_plus_delta * (epsilon + f) * (epsilon_one_plus_delta + t + (delta * t_next)),
        (epsilon_plus_one_delta + h_1 + (h_2 * delta)) * (epsilon_plus_one_delta + h_2 + (h_1_next * delta))
    )
}
```

lopeetall replied:
Combined both maps and combined num + den into single function as suggested

bhgomes commented on plonk-core/src/proof_system/widget/mod.rs:
```diff
@@ -393,6 +425,14 @@ mod test {
+        F: PrimeField,
+    {
+        let values: Vec<_> = (0..n).map(|_| F::rand(&mut OsRng)).collect();
+        MultiSet(values)
```
You should replace this with:

```rust
fn rand_multiset<F>(n: usize) -> MultiSet<F>
where
    F: PrimeField,
{
    let mut rng = OsRng;
    core::iter::from_fn(|| F::rand(&mut rng)).take(n).collect()
}
```

lopeetall replied:
Replaced with your code except using `Some()` around `F::rand(&mut rng)`:

```
    fn rand_multiset<F>(n: usize) -> MultiSet<F>
    where
        F: PrimeField,
    {
        let mut rng = OsRng;
        core::iter::from_fn(|| Some(F::rand(&mut rng)))
            .take(n)
            .collect()
    }
```

bhgomes commented on plonk-core/src/proof_system/widget/lookup.rs:
```diff
@@ -0,0 +1,296 @@
+        zeta: F,
+        lookup_sep: F,
+    ) -> F {
+        // q_lookup(X) * (a(X) + zeta * b(X) + (zeta^2 * c(X)) + (zeta^3 * d(X)
```
The selector is not here.

lopeetall replied:
It is included in the `let a = {...}` closure

bhgomes commented on plonk-core/src/proof_system/widget/lookup.rs:
```diff
@@ -0,0 +1,296 @@
+#[derivative(
+    Clone,
+    Copy(bound = "PC::Commitment: Copy"),
+    Debug(bound = "PC::Commitment: std::fmt::Debug"),
```
Import `core::fmt::Debug` at the top, don't use `std` anywhere. And please change it everywhere else in the library that does this incorrectly.

lopeetall replied:
Changed `std::` to `core::` wherever found

bhgomes commented on plonk-core/src/lookup/lookup_table.rs:
```diff
@@ -0,0 +1,368 @@
+    /// operations that will be done in the plookup table.
+    pub fn insert_special_row(&mut self, a: F, b: F, c: F, d: F) {
+        self.push([a, b, c, d]);
+    }
```
The comment is misleading. I would also just call the name of the method `insert_row`.

lopeetall replied:
Renamed to `insert_row` and rewrote other `insert_{}_row` functions to call it

bhgomes commented on plonk-core/src/lookup/lookup_table.rs:
```diff
@@ -0,0 +1,368 @@
+    /// element must be predetermined to be between -1 and 2 depending on
+    /// the type of table used. If the element does not exist, it will
+    /// return an error.
+    pub fn lookup(&self, a: F, b: F, d: F) -> Result<F, Error> {
```
You should also be able to lookup over any subset of the relation arguments. These tables are not going to just be used for function tables.

CPerezz replied:
I completely agree. Also, why forcing the user to just be able to lookup with 3 values when we can have const_generics and enable basically an arbitrary amount?

lopeetall replied:
Agreed. This will be resolved as a part of #131

bhgomes commented on plonk-core/src/proof_system/linearisation_poly.rs:
```diff
@@ -153,6 +199,8 @@ where
     let omega = domain.group_gen();
     let shifted_z_challenge = *z_challenge * omega;
 
+    // let quot_eval = quot_poly.evaluate(z_challenge);
```
Remove comment?

bhgomes commented on plonk-core/src/proof_system/linearisation_poly.rs:
```diff
@@ -243,14 +337,17 @@ where
+            lookup_evals,
             custom_evals,
         },
+        // quot_eval, // unnecessary?
```
Resolve this

bhgomes commented on plonk-core/src/proof_system/proof.rs:
```diff
@@ -219,15 +286,29 @@ where
+                plonk_verifier_key.lookup.table_3.clone(),
+                plonk_verifier_key.lookup.table_4.clone(),
+            ],
+            &[F::one(), zeta, zeta * zeta, zeta * zeta * zeta],
```
Reduce multiplications here:

```rust
let zeta_sq = zeta.square();
let table_comm = PC::multi_scalar_mul(
    &[
        plonk_verifier_key.lookup.table_1.clone(),
        plonk_verifier_key.lookup.table_2.clone(),
        plonk_verifier_key.lookup.table_3.clone(),
        plonk_verifier_key.lookup.table_4.clone(),
    ],
    &[F::one(), zeta, zeta_sq, zeta_sq * zeta]
);
```

bhgomes commented on plonk-core/src/proof_system/proof.rs:
```diff
@@ -378,6 +494,7 @@ where
     {
         // 5 for each type of gate + 1 for permutations + 4 for each piece of
         // the quotient poly
+        // TODO How much more for lookups?
```
resolve

bhgomes commented on plonk-core/src/proof_system/prover.rs:
```diff
@@ -211,20 +217,135 @@ where
+        // Compute s, as the sorted and concatenated version of f and t
+        let (h_1, h_2) = compressed_t_multiset
+            .sorted_halve(&compressed_f_multiset)
+            .unwrap();
```
Its not actually sorted, see comments in `sorted_halve` implementation.

bhgomes commented on plonk-core/src/proof_system/prover.rs:
```diff
@@ -211,20 +217,135 @@ where
+        let h_2_poly =
+            DensePolynomial::from_coefficients_vec(domain.ifft(&h_2.0));
+
+        // Add blinders to h polynomials
+        // let h_1_poly = Self::add_blinder(&h_1_poly, n, 1);
+        // let h_2_poly = Self::add_blinder(&h_2_poly, n, 1);
```
resolve

lopeetall replied:
We will implement plonkup blinding in a separate PR

bhgomes commented on plonk-core/src/proof_system/prover.rs:
```diff
@@ -247,6 +368,29 @@ where
+            ),
+        );
+
+        // TODO: Find strategy for blinding lookups
+        // Add blinder for lookup permutation poly
+        // z_2_poly = Self::add_blinder(&z_2_poly, n, 2);
```
resolve

lopeetall replied:
We will implement plonkup blinding in a separate PR

bhgomes commented on plonk-core/src/proof_system/prover.rs:
```diff
@@ -352,6 +535,19 @@ where
+        transcript.append(b"h_1_eval", &evaluations.proof.h_1_eval);
+        transcript
+            .append(b"h_1_next_eval", &evaluations.proof.h_1_next_eval);
+        transcript.append(b"h_2_eval", &evaluations.proof.h_2_eval); */
```
resolve

davidnevadoc committed:
Replace `next_u64` for `next_u32`

davidnevadoc committed:
Remove line nit

davidnevadoc committed:
Coherent functions for accessing circuit size

The addition of lookups now requires that the upper bound for the
circuit size be computed as the max between the number of gates and the
lookup table size. To offer a coherent interface to access this value
the following changes have been made:
 - `circuit_size()` function has been removed from `Prover`, `Verifier`
   and `StandardComposer`.
 - `total_size()` has been removed from `Prover`, `Verifier` and in

  `StandardComposer` is now a private method.
 - `circuit_bound()` has been added to `Prover`, `Verifier`and
  `StandardComposer` as the only method available to access this value.

Changes suggested by @bhgomes

davidnevadoc committed:
Add generic types to `util::lc`

davidnevadoc committed:
Remove unused function `random_scalar`

davidnevadoc committed:
Remove unused function `random_scalar`

lopeetall committed:
use cleaner code in pad function

davidnevadoc committed:
Merge branch 'PlookUp' of github.com:rust-zkp/ark-plonk into PlookUp

lopeetall committed:
use more accurate naming and write better docs for splitting function

lopeetall committed:
fix nits

davidnevadoc committed:
Add `lc` for `Multiset`

lopeetall committed:
remove unused functions and their tests

lopeetall committed:
rewrite test for combine_split

lopeetall committed:
use IndexMap instead of HashMap for cleaner combine_split

lopeetall committed:
rewrite comments in combine_split

lopeetall committed:
change publicity and doc string for contains_all

lopeetall committed:
include lookup evals in transcript

lopeetall committed:
clean up lookup table preprocess

lopeetall committed:
reduce multiplications

lopeetall committed:
remove outdated comments on quotient eval

lopeetall committed:
rename and refactor insert row functions

lopeetall committed:
std -> core

lopeetall committed:
rewrite rand_multiset more cleanly

lopeetall committed:
simplify lookup permutation

lopeetall committed:
edit lookup perm poly computation

davidnevadoc committed:
Use `core` instead of `std`

davidnevadoc committed:
Use `rand_core` instead of `rand`

davidnevadoc committed:
Add whitespace nit

lopeetall committed:
change vec capacity to accomodate lookups and fix comment

lopeetall committed:
shorten comments so fmt allows consistency

CPerezz commented on plonk-core/Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "plonk-core"
-version = "0.8.2"
+version = "0.0.0"
```
```suggestion
version = "0.1.0"
```

Minimal version required. Can't be 0.0.0

CPerezz commented on plonk-core/src/constraint_system/lookup.rs:
```diff
@@ -0,0 +1,191 @@
+    /// the largest amount of performance and the minimum circuit-size
+    /// possible. Since it allows the end-user to set every selector coefficient
+    /// as scaling value on the gate eq.
+    pub fn lookup_gate(
```
Would be nice to have some sort of doc here about how the gate eq is evaluated.

davidnevadoc replied:
Scaling the values of the witness with the wire selector values may have resulted in some unintuitive behavior (the witness as a row may not be a row of the lookup table and still be valid) so we removed the wire selectors and got just a normal lookup gate.

CPerezz commented on plonk-core/src/lookup/lookup_table.rs:
```diff
@@ -0,0 +1,365 @@
+/// This struct will be used to determine the outputs of gates within arithmetic
+/// circuits.
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct LookupTable<F>(pub Vec<[F; 4]>)
```
Why can't we use `const_generics` with this? I think would be much more customizable.

davidnevadoc replied:
I agree, but right now we the number of wires we support is fixed at 4 so it wouldn't really add any flexibility. I think this improvement should be added at some point in the future  for both the number of wires and the lookup table width.

CPerezz commented on plonk-core/src/lookup/lookup_table.rs:
```diff
@@ -0,0 +1,365 @@
+        self.insert_row(F::from(a), F::from(b), F::from(c), F::zero());
+    }
+
+    /// Insert a new row for an multiplication operation.
```
```suggestion
    /// Insert a new row for a multiplication operation.
```

CPerezz commented on plonk-core/src/lookup/lookup_table.rs:
```diff
@@ -0,0 +1,365 @@
+            }
+        }
+        table
+    }
```
With all these variants I'm starting to think that maybe the best choice is to make the LookupTable a trait or something similar. And implement `AddLookupTable`, `MulLookupTable` ......


Otherwise this struct is massive and confusing. Containing a lot of different logics and functions for different purposes.

davidnevadoc replied:
Great suggestion! This is a relevant enough change to be made in its own PR (Issue #131 ).

CPerezz commented on plonk-core/src/lookup/lookup_table.rs:
```diff
@@ -0,0 +1,365 @@
+            .is_err());
+    }
+
+    /* FIXME: Check XOR implemenation for fields. */
```
Leftover??

CPerezz replied:
ping @davidnevadoc

CPerezz commented on plonk-core/src/lookup/multiset.rs:
```diff
@@ -0,0 +1,449 @@
+            .flat_map(|item| {
+                // FIXME: Is there a better way to do this in arkworks?
+                let mut bytes = vec![];
+                item.write(&mut bytes).expect("This never fails.");
```
I think maybe `Repr` can help here.

Because https://docs.rs/ark-serialize/latest/ark_serialize/ does not provide alternatives that do not require `Reader`.

CPerezz replied:
ping @davidnevadoc

davidnevadoc replied:
It is AFAIK

CPerezz commented on plonk-core/src/lookup/multiset.rs:
```diff
@@ -0,0 +1,449 @@
+    #[allow(dead_code)]
+    pub(crate) fn contains_all(&self, other: &Self) -> bool {
+        other.0.iter().all(|item| self.contains(item))
+    }
```
If this is only meant to be used in tests, we can maybe add `#[cfg(test)]` instead

CPerezz replied:
ping @davidnevadoc

CPerezz commented on plonk-core/src/lookup/preprocess.rs:
```diff
@@ -0,0 +1,140 @@
+use ark_poly::domain::EvaluationDomain;
+use ark_poly::polynomial::univariate::DensePolynomial;
+
+/// This table will be the preprocessed version of the precomputed table,
+/// T, with arity 4. This structure is passed to the proof alongside the
+/// table of witness values.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct PreprocessedLookupTable<F, PC>
```
Any particular reason why extending Arity would be a big implementation issue? I would really like to understand why we only support 4 columns in the tables.

davidnevadoc replied:
Basically because we only support 4 wire circuits. As I mentioned above we should introduce the flexibility in the table width and number of wires at the same time, but that is a significant change worth its own dedicated PR.

CPerezz commented on plonk-core/src/lookup/witness_table.rs:
```diff
@@ -0,0 +1,172 @@
+/// This table can have any arity. (But for the time other parts of the codebase
+/// force 4-arity)
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct WitnessTable<F>
```
ditto

CPerezz commented on plonk-core/src/proof_system/prover.rs:
```diff
@@ -78,10 +80,9 @@ where
 
-    /// Returns the number of gates in the circuit thet the `Prover` actually
-    /// stores inside.
-    pub fn circuit_size(&self) -> usize {
-        self.cs.circuit_size()
+    /// Returns the smallest power of two needed for the curcuit
+    pub fn circuit_bound(&self) -> usize {
+        self.cs.circuit_bound()
```
I think these functions do not collide. One expresses the required bound while the other returns the current number of gates.

davidnevadoc replied:
`circuit_size` was removed mainly to avoid confusion and prevent misuse, as the value that is normally needed is the smallest power of two that can fit all the circuit (including gates and lookup table).
The number of gates is needed only for a couple of cases and it can be accessed through the standard composer directly.

CPerezz replied:
Ohh ye it's `n`.
Fair enough!

CPerezz commented on plonk-core/src/proof_system/verifier.rs:
```diff
@@ -64,9 +64,9 @@ where
-    /// Returns the number of gates in the circuit.
-    pub fn circuit_size(&self) -> usize {
-        self.cs.circuit_size()
+    /// Returns the smallest power of two needed for the curcuit
+    pub fn circuit_bound(&self) -> usize {
+        self.cs.circuit_bound()
```
ditto

CPerezz reviewed (changes requested):
Brilliant job @lopeetall @davidnevadoc  :tada: :tada: :tada: 

I just have the big concern about: Why not making this through const generics? As far as I've seen in the impl there should not be big barriers to do so. And we could give much more customizable lookups than we do now. 

As it is impossible to build complex lookup expressions with just 4 columns in the lookup table.

davidnevadoc committed:
Spelling nit

Co-authored-by: Carlos Pérez <37264926+CPerezz@users.noreply.github.com>

davidnevadoc committed:
Update version

Co-authored-by: Carlos Pérez <37264926+CPerezz@users.noreply.github.com>

davidnevadoc committed:
Fix benchmarks

Lookup table construction has been moved out of `add_dummy_constraint`
into a separate function `add_dummy_lookup_table`

davidnevadoc committed:
Merge branch 'PlookUp' of github.com:rust-zkp/ark-plonk into PlookUp

davidnevadoc committed:
Move `Multiset::contains` function to test

davidnevadoc committed:
Remove leftover

davidnevadoc committed:
Update benchmarks

LukePearson1 commented on plonk-core/src/lookup/lookup_table.rs:
```diff
@@ -0,0 +1,365 @@
+            for j in lower_bound..upper_bound {
+                table.insert_add_row(i as u64, j as u64, upper_bound as u64);
+            }
+        }
```
Could we just substitute this for:
```rust
pub fn add_table(lower_bound: usize, n: u8) -> LookupTable<F> {
    let mut table = LookupTable::new();
    table.insert_multi_add(lower_bound, n);
    table
}
```

Seems like it's doing the same thing, unless I'm missing something. Same goes for the other table construction fns.

davidnevadoc replied:
Yep, we could reduce this quite a bit. For now I've introduced this change :+1:

LukePearson1 commented on plonk-core/src/lookup/multiset.rs:
```diff
@@ -0,0 +1,448 @@
+
+    /// Compress a vector of multisets into a single multiset using
+    /// a RLC. A random challenge `alpha` needs to be provided. It
+    /// is dervived by hashing the transcript.
```
Typo: `dervived` -> `derived`

LukePearson1 commented on plonk-core/src/lookup/preprocess.rs:
```diff
@@ -0,0 +1,140 @@
+    pub n: u32,
+
+    /// Vector of columns in the preprocessed table containing a
+    /// `MultiSet`, `Commitments` and `DensePolynomial`.
```
Typo: `Commitments` -> `Commitment`

LukePearson1 commented on plonk-core/src/lookup/witness_table.rs:
```diff
@@ -0,0 +1,172 @@
+        right_wire_val: F,
+        fourth_wire_val: F,
+    ) -> Result<(), Error> {
+        // This function is specific for 4-arity
```
Should we put this in the doc comment?

davidnevadoc replied:
I think this may be useful for contributors that may changes to the library but probably not relevant for users, so I think it is fine as a code comment.

LukePearson1 commented on plonk-core/src/lookup/witness_table.rs:
```diff
@@ -0,0 +1,172 @@
+where
+    F: Field,
+{
+    /// Vector containing one `MulitSet` for each wire
```
Typo: `MulitSet` -> `MultiSet`

LukePearson1 commented on plonk-core/src/lookup/witness_table.rs:
```diff
@@ -0,0 +1,172 @@
+use crate::lookup::{LookupTable, MultiSet};
+use ark_ff::Field;
+
+/// This witness table contains quieries to a lookup table for lookup gates.
```
Typo: `quieries` -> `queries`

LukePearson1 commented on plonk-core/src/lookup/witness_table.rs:
```diff
@@ -0,0 +1,172 @@
+where
+    F: Field,
+{
+    /// Initialses empty witness table of arity 4
```
Typo: `Initialses` -> `Initialises`

LukePearson1 commented on plonk-core/src/lookup/witness_table.rs:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    /// This allows the witness table to be filled directly without
+    /// taking any vaules, or the the results, from the lookup table.
```
Typos: `vaules` -> `values`
`the the results` -> `the results`

LukePearson1 commented on plonk-core/src/lookup/witness_table.rs:
```diff
@@ -0,0 +1,172 @@
+
+    /// This allows the witness table to be filled directly without
+    /// taking any vaules, or the the results, from the lookup table.
+    /// If the values do no exists in the lookup table, then the proof
```
Typo: `do no exists` -> `do not exist`

LukePearson1 reviewed (commented):
Just went over the lookup module, pointed out a few nits for redundancy and some typos.

davidnevadoc committed:
Fix typos

davidnevadoc committed:
Nits `lookup_table`

Minor changes to improve code consistency

davidnevadoc committed:
Remove wire selectors from lookup gate

davidnevadoc committed:
Remove leftover

LukePearson1 reviewed (approved):
After several rounds of review, this PR is ready for merge. Any required changes are placed into other issues.
